### PR TITLE
use size_t for buffers in libcurl internal impl

### PR DESCRIPTION
**Pull request by vhvb1989**
fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/2316

### CODE REVIEW

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 1]
```diff
@@ -651,12 +651,12 @@ void CurlSession::ReadStatusLineAndHeadersFromRawResponse(
     bool reuseInternalBuffer)
 {
   auto parser = ResponseBufferParser();
-  auto bufferSize = int64_t();
+  auto bufferSize = int32_t();
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/src/http/curl/curl.cpp:654]: why not `size_t`?

File: sdk/core/azure-core/src/http/curl/curl_connection_private.hpp [hunk 2]
```diff
@@ -20,8 +20,8 @@ namespace Azure { namespace Core { namespace Http {
   namespace _detail {
     // libcurl CURL_MAX_WRITE_SIZE is 64k. Using same value for default uploading chunk size.
     // This can be customizable in the HttpRequest
-    constexpr static int64_t DefaultUploadChunkSize = 1024 * 64;
-    constexpr static auto DefaultLibcurlReaderSize = 1024;
+    constexpr static int32_t DefaultUploadChunkSize = 1024 * 64;
+    constexpr static int32_t DefaultLibcurlReaderSize = 1024;
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/src/http/curl/curl_connection_private.hpp:24]: I feel these should be `size_t`

File: sdk/core/azure-core/src/http/curl/curl_session_private.hpp [hunk 3]
```diff
@@ -238,16 +238,20 @@ namespace Azure { namespace Core { namespace Http {
...
     /**
      * @brief Control field to handle the number of bytes containing relevant data within the
      * internal buffer. This is because internal buffer can be set to be size N but after writing
      * from wire into it, it can be holding less then N bytes.
      *
+     * @note Field will never be greater than `int32_t` because the size of the inner buffer has
+     * type `int32_t`. See m_innerBufferSize.
      */
-    int64_t m_innerBufferSize = _detail::DefaultLibcurlReaderSize;
+    int32_t m_innerBufferSize = _detail::DefaultLibcurlReaderSize;
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/src/http/curl/curl_session_private.hpp:254]: `size_t`?

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:654]: The internal buffer used by our libcurl implementation to parse the response from the wire uses an int32_t buffer.
Also, the current logic is to use a negative index to indicate that the internal buffer has been all parsed or does not contain more data from the wire.

updating to use size_t will require to create a new state to control this case.

Also, would you expect to ever be able to pull more than int32_t bytes from the wire in a single read from a socket?  The current implementation tries to pull  only 1024 bytes from the socket for the internal buffer

File: sdk/core/azure-core/src/http/curl/curl_session_private.hpp (hunk 3 above)
Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl_session_private.hpp:255]: The internal buffer might even be smaller.  We use it just to pull the status line and headers from the payload in chunks of 1024 bytes.

File: sdk/core/azure-core/src/http/curl/curl.cpp (hunk 1 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/src/http/curl/curl.cpp:654]: I'm fine with `int32_t`.
The problem is it violates our guideline `DO Represent memory buffer sizes with size_t or ptrdiff_t as appropriate for the environment.`

Comment (vhvb1989) [reply sdk/core/azure-core/src/http/curl/curl.cpp:654]: I will ask if we can have an exception to that guideline for when the memory buffer is a well known size and will never grow.

File: sdk/core/azure-core/src/http/curl/curl.cpp [hunk 4]
```diff
@@ -576,7 +576,7 @@ void CurlSession::ParseChunkSize(Context const& context)
   // Move to after chunk size
   for (bool keepPolling = true; keepPolling;)
   {
-    for (int64_t index = this->m_bodyStartInBuffer, i = 0; index < this->m_innerBufferSize;
+    for (int32_t index = this->m_bodyStartInBuffer, i = 0; index < this->m_innerBufferSize;
          index++, i++)
```
Comment (antkmsft) [inline sdk/core/azure-core/src/http/curl/curl.cpp:580]: Mixing `index` and `i` probably doesn't read very well. Rename `i` to `iteration` and `index` to `offset`?

**Review by antkmsft — APPROVED** [review]

**Comment by vhvb1989** [comment]
Not to be considered for 1.0  but for 1.1 instead.  
And team wants to follow guidelines and use size_t

**Comment by vhvb1989** [comment]
Updated to follow guideline and use size_t for buffers

**Pull request merged**